카테고리를 통한 중고물품 정렬 및 검색 구현

junshock5 opened the pull request:
ProductSearch 
- Controller 
- Mapper 
- Service  

물품조회 및 정렬 api 개발

junshock5 committed:
ProductSearch Controller, Mapper, Service 생성 및  물품조회 api 개발

junshock5 committed:
REST API URL 패턴 적용

junshock5 committed:
product categoryId를 product Status Enum의 int형으로 set하게 변경
insert 시 categoryId 칼럼 추가
category Insert 시 categoryId 칼럼 추가

junshock5 committed:
검색시 정렬되게 Mybatis case when 동적쿼리 추가
	카테고리별, 판매순, 최신순, 오래된순, 낮은가격순, 높은 가격순, 평점순
	CATEGORIES, SALESCOUNTS, NEWEST, OLDEST, HIGHPRICE, LOWPRICE, GRADE
CategoryDTO 정렬 상태값 SortStatus 변수 추가

junshock5 committed:
commit시 실수로 들어간 문자 삭제

f-lab-dev commented on src/main/java/com/market/server/controller/CategoryController.java:
```diff
@@ -40,7 +40,7 @@ public void registerCategory(String accountId, @RequestBody CategoryDTO category
                                @PathVariable(name = "categoryId") int categoryId,
                                @RequestBody CategoryRequest categoryRequest) {
-        CategoryDTO categoryDTO = new CategoryDTO(categoryId,categoryRequest.getName());
+        CategoryDTO categoryDTO = new CategoryDTO(categoryId,categoryRequest.getName(),null);
```
여기도 `,` 후에 띄어쓰기가 빠져있습니다~

f-lab-dev commented on src/main/resources/mybatis/mapper/productSearchMapper.xml:
```diff
@@ -0,0 +1,47 @@
+            <when test="categoryDTO.sortStatus.toString() == 'LOWPRICE'">
+                ORDER BY #{productDTO.price} ASC
+            </when>
+            <when test="categoryDTO.sortStatus.toString() == 'SALESCOUNTS'">
```
`else`같은 조건은 없을까요?

junshock5 replied:
otherwise가 있습니다.
참고: https://mybatis.org/mybatis-3/ko/dynamic-sql.html

f-lab-dev commented on src/main/resources/mybatis/mapper/productSearchMapper.xml:
```diff
@@ -0,0 +1,47 @@
+            </when>
+            <when test="categoryDTO.sortStatus.toString() == 'GRADE'">
+            </when>
+        </choose>
```
페이징이 없어서 제품이 엄청 많이 검색될 우려가 있습니다

f-lab-dev commented on src/main/java/com/market/server/controller/ProductSearchController.java:
```diff
@@ -0,0 +1,59 @@
+
+    private final ProductSearchServiceImpl productSearchService;
+
+    public ProductSearchController(ProductSearchServiceImpl productSearchService) {
```
lombok을 썼으니 `@RequiredArgConstructor`로 대체하면 어떨까요?

junshock5 replied:
선언된 모든 final필드가 포함된 생성자를 생성해 준다.
(final이 없으면 생성자에 포함되지 않음)

junshock5 committed:
상품 검색시 제품이 많을시 서버 부하를 예방하기 위해 페이징 적용

검색 쿼리에 limit값과 offset 값을 mybatis 쿼리에 연동
CategoryDTO에 searchCount, pagingStartOffset 변수 추가
FINAL 변수 생성자에서 @RequiredArgsConstructor 사용
CategoryController 파라미터 띄어쓰기 추가

f-lab-dev commented on src/main/java/com/market/server/dto/ProductDTO.java:
```diff
@@ -12,7 +12,7 @@
 public class ProductDTO {
     public enum Status {
-        NEW, OLD, ECT
+        NULL, NEW, OLD, ECT
```
보통 NULL이라는 것을 잘 쓰지 않는데 어떤 의미로 null을 쓰셨나요?

junshock5 replied:
유저가 제품 등록시에 상태를 선택하지 않았을때 NULL로 설정하려 했습니다.

f-lab-dev reviewed (commented):
커밋메세지를 보시면 너무 복잡합니다. 커밋메세지도 가독성이 좋도록 예쁘게 적어주세요.

좋은 커밋 메세지 작성하는법 등으로 검색해보시면 좋겠네요

junshock5 commented:
1) 제목과 본문사이 한칸 띄어쓰기
2) 어떻게보다 무엇을, 왜에 맞춰 작성하기
2가지를 6beeb54 에 적용시키겠습니다.

좋은 커밋 메시지 작성법 참고:
https://meetup.toast.com/posts/106

junshock5 committed:
중고 물품 검색시 정렬 구현

카테고리별, 날짜 오르내림차순, 가격별, 유저 등급별 정렬 구현
물품 등록시 가격, 카테고리 정보 등록안할시 예외 발생되게 변경
CategoryDTO 정렬 상태값 중 SALESCOUNTS 삭제

junshock5 committed:
#27 중고 물품 검색시 정렬 구현

카테고리별, 날짜 오르내림차순, 가격별, 유저 등급별 정렬 구현
물품 등록시 가격, 카테고리 정보 등록안할시 예외 발생되게 변경
CategoryDTO 정렬 상태값 중 SALESCOUNTS 삭제

f-lab-dev commented:
커밋메세지에 이슈번호도 같이 넣어주면 좋겠네요 #3 이런식으로요~

junshock5 committed:
Merge branch 'feature/27' of https://github.com/f-lab-edu/used-market-server into feature/27

junshock5 committed:
Merge branch 'develop' into feature/27

junshock5 committed:
Merge remote-tracking branch 'origin/feature/25' into feature/27

f-lab-dev commented on src/main/java/com/market/server/service/Impl/ProductServiceImpl.java:
```diff
@@ -25,7 +26,15 @@ public void register(String id, ProductDTO productDTO) {
-        productDTO.setCategoryId(productDTO.getStatus().ordinal());
+
+        if (productDTO.getStatus().ordinal() == 0 || productDTO.getStatus() == null)
+            throw new RuntimeException("register ERROR! 상품 categoryId를 확인해주세요\n" + "categoryId : " + productDTO.getStatus().ordinal());
```
`ordinal`이 어떨 때 쓰는것인지 확인해보시고 이것을 ID값으로 사용하는게 적절한 사용처인지 고민해보세요~

junshock5 replied:
ordinal() 메소드는 해당 열거체 상수가 열거체 정의에서 정의된 순서(0부터 시작)를 반환합니다.
상수값 자체가 아니므로 상수로 비교하는게 더 명시적으로 생각되서 변경하겠습니다.

f-lab-dev replied:
카톡으로 얘기한것처럼 DB에서 카테고리를 관리하는게 더 낫겠네요~
자바에서 관리하면 카테고리가 늘어날때마다 재컴파일해야하니까요

junshock5 committed:
물품 카테고리 번호를 DB에서 관리하게 변경

f-lab-dev commented on src/main/java/com/market/server/controller/ProductController.java:
```diff
@@ -58,7 +58,7 @@ public ProductResponse myProductInfo(String accountId) {
     public void updateProducts(String accountId,
                                @PathVariable(name = "productId") int productId,
-                               @RequestBody ProductRequest productRequest) {
+                               @RequestBody ProductRequest PR) {
```
작명규칙 지켜주세요~

junshock5 replied:
feature/18 브랜치에서 변경하였습니다~

junshock5 committed:
Merge branch 'develop' into feature/27
